### PR TITLE
Convert errors in `@solana/transactions` to coded exceptions

### PR DESCRIPTION
**Pull request by steveluscher**
Addresses #2118.

### CODE REVIEW

**Comment by steveluscher** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2207?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* **#2208** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2208?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2207** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2207?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#2206** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2206?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/2207?utm_source=stack-comment).

File: packages/errors/src/messages.ts [hunk 1]
```diff
@@ -35,4 +44,23 @@ export const SolanaErrorMessages: Readonly<{
     [SOLANA_ERROR__TRANSACTION_SIGNATURE_NOT_COMPUTABLE]:
         "Could not determine this transaction's signature. Make sure that the transaction has " +
         'been signed by its fee payer.',
+    [SOLANA_ERROR__TRANSACTION_INVOKED_PROGRAMS_CANNOT_PAY_FEES]:
+        'This transaction includes an address (`$programAddress`) which is both ' +
+        'invoked and set as the fee payer. Program addresses may not pay fees',
```
Comment (steveluscher) [inline packages/errors/src/messages.ts:49]: I'm not 100% sure if this error should be this, or:


```suggestion
    [SOLANA_ERROR__TRANSACTION_PROGRAMS_CANNOT_PAY_FEES]:
        'This transaction includes a program (`$programAddress`) which is set as ' +
        'the fee payer. Program addresses may not pay fees',
```

ie. does it matter that it's invoked, or is it that programs can't pay fees, period? cc/ @joncinque

File: packages/transactions/src/decompile-transaction.ts [hunk 2]
```diff
@@ -84,9 +91,13 @@ function getAddressLookupMetas(
...
         const highestIndex = Math.max(...lookup.readableIndices, ...lookup.writableIndices);
         if (highestIndex >= addresses.length) {
-            // TODO coded error
-            throw new Error(
-                `Cannot look up index ${highestIndex} in lookup table [${lookup.lookupTableAddress}]. The lookup table may have been extended since the addresses provided were retrieved.`,
+            throw new SolanaError(
+                SOLANA_ERROR__TRANSACTION_FAILED_TO_DECOMPILE_ADDRESS_LOOKUP_TABLE_INDEX_OUT_OF_RANGE,
+                {
+                    highestKnownIndex: addresses.length - 1,
+                    highestRequestedIndex: highestIndex,
```
Comment (steveluscher) [inline packages/transactions/src/decompile-transaction.ts:98]: Modified this to include the highest index available.

File: packages/errors/src/messages.ts [hunk 3]
```diff
@@ -31,8 +40,27 @@ export const SolanaErrorMessages: Readonly<{
         'The $argumentLabel argument to the `$methodName` RPC method$optionalPathLabel was ' +
         '`$value`. This number is unsafe for use with the Solana JSON-RPC because it exceeds ' +
         '`Number.MAX_SAFE_INTEGER`.',
+    [SOLANA_ERROR__TRANSACTION_EXPECTED_BLOCKHASH_LIFETIME]: 'Transaction does not have a blockhash lifetime',
+    [SOLANA_ERROR__TRANSACTION_EXPECTED_NONCE_LIFETIME]: 'Transaction is not a durable nonce transaction',
+    [SOLANA_ERROR__TRANSACTION_FAILED_TO_DECOMPILE_ADDRESS_LOOKUP_TABLE_CONTENTS_MISSING]:
+        'Contents of these address lookup tables unknown: $lookupTableAddresses',
```
Comment (steveluscher) [inline packages/errors/src/messages.ts:46]: Changed this.

File: packages/errors/src/codes.ts [hunk 4]
```diff
@@ -13,6 +13,15 @@ export const SOLANA_ERROR__INVALID_KEYPAIR_BYTES = 4 as const;
...
 export const SOLANA_ERROR__NONCE_ACCOUNT_NOT_FOUND = 7 as const;
+export const SOLANA_ERROR__TRANSACTION_INVOKED_PROGRAMS_CANNOT_PAY_FEES = 99901 as const;
+export const SOLANA_ERROR__TRANSACTION_INVOKED_PROGRAMS_MUST_NOT_BE_WRITABLE = 99902 as const;
+export const SOLANA_ERROR__TRANSACTION_EXPECTED_BLOCKHASH_LIFETIME = 99903 as const;
+export const SOLANA_ERROR__TRANSACTION_EXPECTED_NONCE_LIFETIME = 99904 as const;
+export const SOLANA_ERROR__TRANSACTION_VERSION_NUMBER_OUT_OF_RANGE = 99905 as const;
+export const SOLANA_ERROR__TRANSACTION_FAILED_TO_DECOMPILE_ADDRESS_LOOKUP_TABLE_CONTENTS_MISSING = 99906 as const;
+export const SOLANA_ERROR__TRANSACTION_FAILED_TO_DECOMPILE_ADDRESS_LOOKUP_TABLE_INDEX_OUT_OF_RANGE = 99907 as const;
+export const SOLANA_ERROR__TRANSACTION_FAILED_TO_DECOMPILE_INSTRUCTION_PROGRAM_ADDRESS_NOT_FOUND = 99908 as const;
+export const SOLANA_ERROR__TRANSACTION_FAILED_TO_DECOMPILE_FEE_PAYER_MISSING = 99909 as const;
```
Comment (lorisleiva) [inline packages/errors/src/codes.ts:24]: I don't feel too strongly about this but how do you guys feel about double-underscoring after the prefix so it's clear that we are creating a "group" of error codes?

Cc @buffalojoec @steveluscher @mcintyre94 

```ts
export const SOLANA_ERROR__TRANSACTION__INVOKED_PROGRAMS_CANNOT_PAY_FEES = 99901 as const;
export const SOLANA_ERROR__TRANSACTION__INVOKED_PROGRAMS_MUST_NOT_BE_WRITABLE = 99902 as const;
export const SOLANA_ERROR__TRANSACTION__EXPECTED_BLOCKHASH_LIFETIME = 99903 as const;
export const SOLANA_ERROR__TRANSACTION__EXPECTED_NONCE_LIFETIME = 99904 as const;
export const SOLANA_ERROR__TRANSACTION__VERSION_NUMBER_OUT_OF_RANGE = 99905 as const;
export const SOLANA_ERROR__TRANSACTION__FAILED_TO_DECOMPILE_ADDRESS_LOOKUP_TABLE_CONTENTS_MISSING = 99906 as const;
export const SOLANA_ERROR__TRANSACTION__FAILED_TO_DECOMPILE_ADDRESS_LOOKUP_TABLE_INDEX_OUT_OF_RANGE = 99907 as const;
export const SOLANA_ERROR__TRANSACTION__FAILED_TO_DECOMPILE_INSTRUCTION_PROGRAM_ADDRESS_NOT_FOUND = 99908 as const;
export const SOLANA_ERROR__TRANSACTION__FAILED_TO_DECOMPILE_FEE_PAYER_MISSING = 99909 as const;
```

Comment (buffalojoec) [reply packages/errors/src/codes.ts:24]: I also don't feel too strongly here but the double underscore is a bit more readable.

Comment (steveluscher) [reply packages/errors/src/codes.ts:24]: Looks like Next.js' build system inlines the codes just fine, so however long we want to make these names is fine by me!

**Pull request merged**